src/main/java/io/cdap/plugin/snowflake/source/batch/SnowflakeMapToRecordTransformer.java

```diff
@@ -0,0 +1,127 @@
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Transforms Snowflake row into {@link StructuredRecord}.
+ */
+public class SnowflakeMapToRecordTransformer {
+  private static final Logger LOG = LoggerFactory.getLogger(SnowflakeMapToRecordTransformer.class);
+
+  private final Schema schema;
+
+  public SnowflakeMapToRecordTransformer(Schema schema) {
+    this.schema = schema;
+  }
+
+  public StructuredRecord transform(Map<String, String> row) {
+    return getStructuredRecord(row, schema);
+  }
+
+  private StructuredRecord getStructuredRecord(Map<String, String> row, Schema schema) {
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    row.entrySet().stream()
+      .filter(entry -> schema.getField(entry.getKey()) != null) // filter absent fields in the schema
+      .forEach(entry -> builder.set(
+        entry.getKey(),
+        convertValue(entry.getKey(), entry.getValue(), schema.getField(entry.getKey()).getSchema())));
+    return builder.build();
+  }
+
+  private Object convertValue(String fieldName, String value, Schema fieldSchema) {
+    if (fieldSchema.isNullable()) {
+      return convertValue(fieldName, value, fieldSchema.getNonNullable());
+    }
+
+    Schema.Type fieldSchemaType = fieldSchema.getType();
+
+    // empty string is considered null in csv
+    if (value == null || (value instanceof String && Strings.isNullOrEmpty((String) value))) {
+      return null;
+    }
+
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (logicalType != null) {
+      switch (logicalType) {
+        case DATE:
+          // date will be in yyyy-mm-dd format
+          return Math.toIntExact(LocalDate.parse(String.valueOf(value)).toEpochDay());
+        case TIMESTAMP_MICROS:
+          Instant instant = DateTimeFormatter.ISO_OFFSET_DATE_TIME.parse(String.valueOf(value), Instant::from);
+          return TimeUnit.MILLISECONDS.toMicros(instant.toEpochMilli());
+        case TIME_MICROS:
+          return TimeUnit.NANOSECONDS.toMicros(LocalTime.parse(String.valueOf(value)).toNanoOfDay());
+        case DECIMAL:
+          return new BigDecimal(String.valueOf(value)).setScale(fieldSchema.getScale()).unscaledValue().toByteArray();
+        default:
+          throw new IllegalArgumentException(
+            String.format("Field '%s' is of unsupported type '%s'", fieldName, logicalType.getToken()));
+      }
+    }
+
+    switch (fieldSchemaType) {
+      case NULL:
+        return null;
+      case BYTES:
+        // decode hex value
+        return hexStringToByteArray(value);
+      case BOOLEAN:
+        return Boolean.parseBoolean(castValue(value, fieldName, String.class));
+      case DOUBLE:
+        return Double.parseDouble(castValue(value, fieldName, String.class));
+      case STRING:
+        return value;
+    }
+
+    throw new UnexpectedFormatException(
```

Comment:
snowflake does not support complex types like array, structured record etc?

Reply:
The usual way to store complex types in to use VARIANT. The VARIANT is a type which can store any semi-structured data like json/xml etc. 
In this plugin we map VARIANT values to string. Since their structure is unknown like some xml data.

Reply:
It has logical types object and array. **But** in reality they are strings with jsons. Even if you try to query them via jdbc and ask the type it will return VARCHAR.